### PR TITLE
[SWDEV-304151] clear program cache when tuning is done

### PR DESCRIPTION
**Pull request by JehandadKhan**
Addresses #1223 by forcing the destruction of all the interim hip program objects created during the tuning loop

Also removes tuning for ConvAsm1x1 fused solver

### CODE REVIEW

File: src/hip/handlehip.cpp [hunk 1]
```diff
@@ -447,6 +447,13 @@ void Handle::AddProgram(Program prog,
     this->impl->cache.AddProgram(prog, program_name, params);
 }
 
+void Handle::ClearProgram(const std::string& program_name, const std::string& params)
```
Comment (junliume) [inline src/hip/handlehip.cpp:450]:
```
[2021-10-11T21:09:20.706Z] /var/jenkins/workspace/MLLibs_MIOpen_jd_file_leak/src/hip/handlehip.cpp:450:14: error: method 'ClearProgram' can be made const [readability-make-member-function-const,-warnings-as-errors]

[2021-10-11T21:09:20.706Z] void Handle::ClearProgram(const std::string& program_name, const std::string& params)

[2021-10-11T21:09:20.706Z]              ^

[2021-10-11T21:09:20.706Z]                                                                                       const

File: src/solver/conv_asm_1x1u_bias_activ.cpp [hunk 2]
```diff
@@ -104,8 +104,10 @@ ConvBiasActivAsm1x1U::Search(const ConvolutionContext& context, const AnyInvokeP
     tensors.bias    = bias_buf.get();
 
     const auto fused_invoke_ctx = conv::FusedDataInvokeParams(tensors, nullptr, 0);
```
Comment (junliume) [inline src/solver/conv_asm_1x1u_bias_activ.cpp:106]: maybe delete this line?
```
[2021-10-12T17:28:47.168Z] /var/jenkins/workspace/MLLibs_MIOpen_jd_file_leak/src/solver/conv_asm_1x1u_bias_activ.cpp:106:33: warning: Variable 'fused_invoke_ctx' is assigned a value that is never used. [unreadVariable]
[2021-10-12T17:28:47.168Z]     const auto fused_invoke_ctx = conv::FusedDataInvokeParams(tensors, nullptr, 0);

File: src/hip/handlehip.cpp [hunk 3]
```diff
@@ -447,6 +447,13 @@ void Handle::AddProgram(Program prog,
     this->impl->cache.AddProgram(prog, program_name, params);
 }
 
+void Handle::ClearProgram(const std::string& program_name, const std::string& params) const
+{
+    this->impl->cache.ClearProgram(program_name, params);
+}
+
+void Handle::ClearProgram() const { this->impl->cache.ClearProgram(); }
```
Comment (pfultz2) [inline src/hip/handlehip.cpp:455]: This should probably be named `ClearPrograms` to make it clearer that it removes all programs.

File: src/nogpu/handle.cpp [hunk 4]
```diff
@@ -147,6 +147,12 @@ void Handle::ClearKernels(const std::string& algorithm, const std::string& netwo
 {
     this->impl->cache.ClearKernels(algorithm, network_config);
 }
+void Handle::ClearProgram(const std::string& program_name, const std::string& params) const
```
Comment (pfultz2) [inline src/nogpu/handle.cpp:150]: Whats the purpose of this function? It doesnt seemed to be used anywhere.

File: src/solver/conv_asm_1x1u_bias_activ.cpp [hunk 5]
```diff
@@ -103,9 +103,10 @@ ConvBiasActivAsm1x1U::Search(const ConvolutionContext& context, const AnyInvokeP
     tensors.outDesc = context.conv_problem.GetOut();
     tensors.bias    = bias_buf.get();
 
-    const auto fused_invoke_ctx = conv::FusedDataInvokeParams(tensors, nullptr, 0);
-
-    return GenericSearch(*this, cba_context, fused_invoke_ctx);
+    PerformanceConfigConvBiasActivAsm1x1U pp;
+    pp.HeuristicInit(context);
+    return pp;
+    // return GenericSearch(*this, cba_context, fused_invoke_ctx);
```
Comment (atamazov) [inline src/solver/conv_asm_1x1u_bias_activ.cpp:109]: Why this change is necessary?

**Comment by junliume** [comment]
@JehandadKhan @atamazov this might have caused [SWDEV-316216](https://ontrack-internal.amd.com/browse/SWDEV-316216), which blocks 5.0. Any risk assessment about reverting this one?

**Comment by atamazov** [comment]
> @JehandadKhan @atamazov this might have caused [SWDEV-316216](https://ontrack-internal.amd.com/browse/SWDEV-316216), which blocks 5.0. Any risk assessment about reverting this one?

The likely reason of the issue is recent patch in ROCClr. More info at https://ontrack-internal.amd.com/browse/SWDEV-304151?focusedCommentId=8279176&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-8279176.

Reverting PR #1221 would lead to reappearing of SWDEV-304151 as soon as the ROCClr issue is fixed.

File: src/include/miopen/generic_search.hpp [hunk 6]
```diff
@@ -466,6 +466,8 @@ auto GenericSearch(const Solver s, const Context& context_, const AnyInvokeParam
 
     MIOPEN_LOG_W("Done: " << n_runs_total << '/' << n_failed << '/' << n_runs_total << ", best #"
                           << n_best << ' ' << best_time << ' ' << best_config);
+
+    profile_h.ClearProgram();
```
Comment (atamazov) [inline src/include/miopen/generic_search.hpp:470]: Clearing all Program objects invalidates all Kernel objects, including those that have been built _before_ tuning. This would lead to failures of applications that leverage "on-the-fly" tuning by means of `MIOPEN_FIND_ENFORCE`.

**Review by atamazov — COMMENTED** [review]
:bug:

File: src/include/miopen/generic_search.hpp (hunk 6 above)
Comment (atamazov) [reply src/include/miopen/generic_search.hpp:470]: I am working on the patch.
